### PR TITLE
Interface clean up

### PR DESCRIPTION
**Pull request by bitwiseman**
There were a number of tweaks to the interface that seemed sensible to me.  If you want I can break them out into separate pull requests, but they can get a tedious. 

Overview:
 * dao.push() should not log status. It should either succeed or throw, nothing else.   Status logging (if any) should be handled by the wrapper, notifier, or stream. 
 * dao.getDescription() - instead of handing out the port and host, what we really want is to let the dao give us a description of its settings in a format that is meaningful to it. 
 * `final` and package private - most of the uses of `protected` don't need to be there.  Many fields are assigned once and should communicate that by being `final`. 
 * disallow null where possible - there was a path for getDao() that checked the host value and returned null.  That is the responsibility of the constructor and it should throw.

### CODE REVIEW

File: src/main/java/jenkins/plugins/logstash/LogstashInstallation.java [hunk 1]
```diff
@@ -95,13 +95,9 @@ public String getDisplayName() {
...
 
     /**
-     * @see IndexerDaoFactory#getInstance(IndexerType, String, int, String, String)
+     * @see IndexerDaoFactory#getInstance(IndexerType, String, int, String, String, String)
      */
     public LogstashIndexerDao getIndexerDao() throws InstantiationException {
-      if (StringUtils.isEmpty(host)) {
-        return null;
-      }
-
```
Comment (bitwiseman) [inline src/main/java/jenkins/plugins/logstash/LogstashInstallation.java]: Constructor should handle this case, not this method.

File: src/main/java/jenkins/plugins/logstash/LogstashNotifier.java [hunk 2]
```diff
@@ -55,21 +55,13 @@
...
   public int maxLines;
   public boolean failBuild;
 
   @DataBoundConstructor
   public LogstashNotifier(int maxLines, boolean failBuild) {
-    this(maxLines, failBuild, Jenkins.getInstance().getRootUrl());
-  }
-
-  // Constructor for unit tests
-  protected LogstashNotifier(int maxLines, boolean failBuild, String jenkinsUrl) {
```
Comment (bitwiseman) [inline src/main/java/jenkins/plugins/logstash/LogstashNotifier.java]: Instead of having a constructor just for test have an override.

File: src/main/java/jenkins/plugins/logstash/LogstashNotifier.java [hunk 3]
```diff
@@ -83,11 +75,6 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
       return !failBuild;
     }
 
-    if (dao == null) {
-      listener.getLogger().println("[logstash-plugin]: Host name not specified. Unable to send log data.");
-      return !failBuild;
-    }
-
```
Comment (bitwiseman) [inline src/main/java/jenkins/plugins/logstash/LogstashNotifier.java]: dao should now never be null here.

File: src/test/java/jenkins/plugins/logstash/LogstashBuildWrapperTest.java [hunk 4]
```diff
@@ -38,12 +38,16 @@
     }
 
     @Override
-    protected LogstashIndexerDao getDao() {
+    LogstashIndexerDao getDao() throws InstantiationException {
+      if (dao == null) {
+        throw new InstantiationException("DoaTestInstantiationException");
+      }
```
Comment (bitwiseman) [inline src/test/java/jenkins/plugins/logstash/LogstashBuildWrapperTest.java]: Test the instantiation error case.

**Comment by R-Gerard** [comment]
I'm going out of town. I'll review this sometime later next week.

**Comment by bitwiseman** [comment]
Sure, no hurry.  It's an api level change, but doesn't change the customer level features.

File: src/main/java/jenkins/plugins/logstash/LogstashInstallation.java (hunk 1 above)
Comment (R-Gerard) [inline src/main/java/jenkins/plugins/logstash/LogstashInstallation.java]: This was handling an edge case where the plugin has been installed but not yet configured:

1. Install the plugin, but don't edit the global configurations
2. Add the plugin to a job (either the post-build action or the buildwrapper)
3. Run the job

Here's what happens with the post-build action:

```
ERROR: Publisher jenkins.plugins.logstash.LogstashNotifier aborted due to exception
java.lang.NullPointerException
	at jenkins.plugins.logstash.LogstashInstallation$Descriptor.getIndexerDao(LogstashInstallation.java:101)
	at jenkins.plugins.logstash.LogstashNotifier.getDao(LogstashNotifier.java:109)
	at jenkins.plugins.logstash.LogstashNotifier.perform(LogstashNotifier.java:71)
	at hudson.tasks.BuildStepMonitor$3.perform(BuildStepMonitor.java:45)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:772)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:736)
	at hudson.model.Build$BuildExecution.post2(Build.java:183)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:685)
	at hudson.model.Run.execute(Run.java:1757)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:88)
	at hudson.model.Executor.run(Executor.java:234)
```

With the buildwrapper it fails catastrophically (this is from the server log, not the build log):

```
SEVERE: #5 failed to build and we don't even have a listener
java.lang.NullPointerException
        at jenkins.plugins.logstash.LogstashInstallation$Descriptor.getIndexerDao(LogstashInstallation.java:101)
        at jenkins.plugins.logstash.LogstashBuildWrapper.getDao(LogstashBuildWrapper.java:95)
        at jenkins.plugins.logstash.LogstashBuildWrapper.decorateLogger(LogstashBuildWrapper.java:77)
        at hudson.model.Run.execute(Run.java:1711)
        at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
        at hudson.model.ResourceController.execute(ResourceController.java:88)
        at hudson.model.Executor.run(Executor.java:234)
```

**Comment by R-Gerard** [comment]
Thanks for the submission. I like the changes you made to the interface. Just amend e2c7c4c with a fix for the NPE and I think we're good.

File: src/main/java/jenkins/plugins/logstash/LogstashInstallation.java [hunk 5]
```diff
@@ -95,14 +95,10 @@ public String getDisplayName() {
...
-      return IndexerDaoFactory.getInstance(type, host, port, key, username, password);
+      return IndexerDaoFactory.getInstance(type, host, (port == null ? -1 : port.intValue()), key, username, password);
```
Comment (bitwiseman) [inline src/main/java/jenkins/plugins/logstash/LogstashInstallation.java]: The port is represented as `Integer` in the Descriptor and `int` in the DAO. :cry: This handles the null case.

Comment (R-Gerard) [inline src/main/java/jenkins/plugins/logstash/LogstashInstallation.java]: Just change line 68 to `public Integer port = -1;`

**Comment by bitwiseman** [comment]
> Just change line 68 to public Integer port = -1;

Done.  But also made the IndexerFactory work with `Integer` and added tests for correct handling of `null`.  It should never be null, but if it is we're not going blow up.

**Comment by bitwiseman** [comment]
Updated my change.  FYI, I have another commit lined up after this the further improves error reporting.

**Comment by R-Gerard** [comment]
Thanks again for the improvements.
